src/hotspot/share/services/attachListener.cpp

```diff
@@ -47,10 +48,124 @@
 #include "utilities/debug.hpp"
 #include "utilities/formatBuffer.hpp"
 
+
+// Stream for printing attach operation results.
+// Supports buffered and streaming output for commands which can produce lengthy reply.
+//
+// A platform implementation supports streaming output if it implements AttachOperation::get_reply_writer().
+// Streaming is enabled if the allow_streaming in the constructor is set to true.
+//
+// Initially attachStream works in buffered mode.
+// To switch to the streaming mode attach command handler need to call attachStream::set_result().
+// The method flushes buffered output and consequent printing goes directly to ReplyWriter.
+class attachStream : public bufferedStream {
+  AttachOperation::ReplyWriter* _reply_writer;
+  bool _allow_streaming;
+  jint _result;
+  bool _result_set;
+  bool _result_written;
+  bool _error;
+
+  enum : size_t {
+    INITIAL_BUFFER_SIZE = 1 * M,
+    MAXIMUM_BUFFER_SIZE = 3 * G,
+  };
+
+  bool is_streaming() const {
+    return _result_set && _allow_streaming;
+  }
+
+  void flush_reply() {
+    if (_error) {
+      return;
+    }
+
+    if (!_result_written) {
+      if (_reply_writer->write_reply(_result, this)) {
+        _result_written = true;
+        reset();
+      } else {
+        _error = true;
+        return;
+      }
+    } else {
+      _error = !_reply_writer->write_fully(base(), (int)size());
+      reset();
+    }
+  }
+
+public:
+  attachStream(AttachOperation::ReplyWriter* reply_writer, bool allow_streaming)
+    : bufferedStream(INITIAL_BUFFER_SIZE, MAXIMUM_BUFFER_SIZE),
+      _reply_writer(reply_writer),
+      _allow_streaming(reply_writer == nullptr ? false : allow_streaming),
+      _result(JNI_OK), _result_set(false), _result_written(false),
+      _error(false)
+    {}
+
+  virtual ~attachStream() {}
+
+  void set_result(jint result) {
+    if (!_result_set) {
+      _result = result;
+      _result_set = true;
+      if (_allow_streaming) {
+        // switch to streaming mode
+        flush_reply();
+      }
+    }
+  }
+
+  jint get_result() const {
+    return _result;
+  }
+
+  bufferedStream* get_buffered_stream() {
+    return this;
+  }
+
+  // Called after the operation is completed.
+  // If reply_writer is provided, writes the results.
+  void complete() {
+    flush_reply();
+  }
+
+  virtual void write(const char* str, size_t len) override {
+    if (is_streaming()) {
+      if (!_error) {
+        _error = !_reply_writer->write_fully(str, (int)len);
+      }
```

Comment:
What happens if we're in  `is_streaming()` but the result isn't written? That breaks the protocol. At least add an assert that `_result_written` is set to true.